beacon/beacon.go

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (C) 2015 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+// Package beacon contains an implementation of GA4GH Beacon API (http://ga4gh.org/#/beacon).
+package beacon
+
+import (
+	"encoding/xml"
+	"fmt"
+	"html/template"
+	"net/http"
+	"strconv"
+
+	"github.com/googlegenomics/beacon-go/internal/variants"
+	"google.golang.org/appengine"
+)
+
+const (
+	// beaconAPIVersion the version of the GA4GH Beacon specification the API implements.
+	beaconAPIVersion = "v0.0.1"
+
+	aboutDefaultPath = "/"
+	queryDefaultPath = "/query"
+)
+
+var (
+	aboutTemplate = template.Must(template.ParseFiles("about.xml"))
+)
+
+// Server implements a GA4GH Beacon API (https://github.com/ga4gh-beacon/specification/blob/master/beacon.md) backed
+// by a Google Cloud BigQuery variants table.
+type Server struct {
+	// ProjectID is the GCloud project ID.
+	ProjectID string
+	// TableID is the ID of the allele BigQuery table to query.
+	// Must be provided in the following format: bigquery-project.dataset.table.
+	TableID string
+}
+
+// Export registers the beacon API endpoint with mux.
+func (server *Server) Export(mux *http.ServeMux) {
+	mux.Handle(aboutDefaultPath, forwardOrigin(server.About))
+	mux.Handle(queryDefaultPath, forwardOrigin(server.Query))
+}
+
+// About retrieves all the necessary information on the beacon and the API.
+func (api *Server) About(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		http.Error(w, fmt.Sprintf("HTTP method %s not supported", r.Method), http.StatusBadRequest)
+		return
+	}
+	w.Header().Set("Content-Type", "application/xml")
+
+	info := struct {
```

Comment:
Optional, but for things like this I tend to just use maps

aboutTemplate.Execute(w, map[string]string{
  "APIVersion": apiVersion,
  "TableID": tableID,
})

This may make less sense when you introduce the 'api' package later.

Reply:
done. this will change